system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,146 @@
+    return items[items.length - 1].toUpperCase();
+  }
+
+  downloadsCmd_copyLocation() {
+    let clipboard = Cc["@mozilla.org/widget/clipboardhelper;1"].getService(Ci.nsIClipboardHelper);
+    clipboard.copyString(this.download.source.url);
+  }
+}
+
+this.DownloadsManager = class DownloadsManager {
+  constructor(store) {
+    this._downloadData = null;
+    this._window = null;
+    this._store = null;
+    this._viewableDownloadItems = new Map();
+  }
+
+  formatDownload(element) {
+    const downloadedItem = element.download;
+    let description;
+    if (element.fileType) {
+      // If we have a file type: '1.5 MB — PNG'
+      description = `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`;
+    } else {
+      // If we do not have a file type: '1.5 MB'
+      description = `${element.sizeStrings.stateLabel}`;
+    }
+    return {
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description,
+      referrer: downloadedItem.source.referrer,
+      date_added: downloadedItem.endTime
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      this._viewableDownloadItems.set(downloadedItem.source.url, elem);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  onDownloadRemoved(download) {
+    if (this._viewableDownloadItems.has(download.source.url)) {
+      this._viewableDownloadItems.delete(download.source.url);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  get downloads() {
+    let results = [];
+    for (const url of this._viewableDownloadItems.keys()) {
+      const elem = this._viewableDownloadItems.get(url);
+      // We only want downloads that are less than 36 hours old
+      if (elem && (elem.downloadedSince < RECENT_DOWNLOAD_THRESHOLD)) {
+        const formattedDownloadForHighlights = this.formatDownload(elem);
+        results.push(formattedDownloadForHighlights);
+      }
+    }
+    return results.sort((a, b) => a.date_added < b.date_added);
+  }
+
+  uninit() {
+    if (this._downloadData) {
+      this._downloadData.removeView(this);
+      this._downloadData = null;
+    }
+  }
+
+  onAction(action) {
+    let downloadsCmd;
+    switch (action.type) {
+      case at.COPY_DOWNLOAD_LINK:
+        downloadsCmd = "copyLocation";
+        break;
+      case at.GO_TO_DOWNLOAD_PAGE:
+        downloadsCmd = "openReferrer";
+        break;
+      case at.REMOVE_DOWNLOAD_FILE:
+        downloadsCmd = "delete";
```

Comment:
Looks like it's possible for this to throw if the file was deleted.. although not sure why it doesn't show up as disabled?
```
[Exception... "Component returned failure code: 0x80520006 (NS_ERROR_FILE_TARGET_DOES_NOT_EXIST) [nsIFile.isExecutable]"  nsresult: "0x80520006 (NS_ERROR_FILE_TARGET_DOES_NOT_EXIST)"  location: "JS frame :: resource:///modules/DownloadsCommon.jsm :: openDownloadedFile :: line 423"  data: no]
```
<img width="522" alt="screen shot 2018-04-18 at 2 39 19 pm" src="https://user-images.githubusercontent.com/438537/38959665-59f84976-4316-11e8-9542-c055b6e3151e.png">